test/jdk/java/lang/String/StringRacyConstructor.java

```diff
@@ -0,0 +1,441 @@
+        var helloWorld = "hello world";
+        String helloToo = racyStringConstruction("hell".concat("o"));
+        String o = helloToo.intern();
+        var hello = "hello";
+        assertTrue(validCoder(helloToo), "startsWith: "
+                + ", hell: " + inspectString(helloToo)
+                + ", o: " + inspectString(o)
+                + ", hello: " + inspectString(hello)
+                + ", hello world: " + inspectString(helloWorld));
+    }
+
+    // Check that an empty string with racy construction has a valid coder
+    @Test
+    @EnabledIf("test.java.lang.String.StringRacyConstructor#isCompactStrings")
+    public void racyEmptyString() {
+        var space = racyStringConstruction(" ");
+        var trimmed = space.trim();
+        assertTrue(validCoder(trimmed), "empty string invalid coder"
+                + ", trimmed: " + inspectString(trimmed));
+    }
+
+    // Check that an exception in a user implemented CharSequence doesn't result in
+    // an invalid coder when appended to a StringBuilder
+    @Test
+    @EnabledIf("test.java.lang.String.StringRacyConstructor#isCompactStrings")
+    void charSequenceException() {
+        ThrowingCharSequence bs = new ThrowingCharSequence("A\u2030\uFFFD");
+        var sb = new StringBuilder();
+        try {
+            sb.append(bs);
+            fail("An IllegalArgumentException should have been thrown");
+        } catch (IllegalArgumentException ex) {
+            // ignore expected
+        }
+        assertTrue(validCoder(sb), "invalid coder in StringBuilder");
+    }
+
+    /**
+     * Given a latin-1 String, attempt to create a copy that is
+     * incorrectly encoded as UTF-16.
+     */
+    public static String racyStringConstruction(String original) throws ConcurrentModificationException {
+        if (original.chars().max().getAsInt() >= 256) {
+            throw new IllegalArgumentException(
+                    "Only work with latin-1 Strings");
+        }
+
+        char[] chars = original.toCharArray();
+
+        // In another thread, flip the first character back
+        // and forth between being latin-1 or not
+        Thread thread = new Thread(() -> {
+            while (!Thread.interrupted()) {
+                chars[0] ^= 256;
+            }
+        });
+        thread.start();
+
+        // at the same time call the String constructor,
+        // until we hit the race condition
+        int i = 0;
+        while (true) {
+            i++;
+            String s = new String(chars);
+            if ((s.charAt(0) < 256 && !original.equals(s)) || i > 1_000_000) {
+                thread.interrupt();
+                try {
+                    thread.join();
+                } catch (InterruptedException ie) {
+                    // ignore interrupt
+                }
+                if (i >= 1_000_000) {
+                    System.out.printf("Unable to produce a UTF16 string in %d iterations: %s%n", i, original);
```

Comment:
What I meant with my previous comment is that this leaves a trace on "success", which is kind of unusual.
There might be good reasons to do so, but they are not apparent to me.